### PR TITLE
Add `quorum_active` flag to cluster health and waitForQuorumActive

### PR DESCRIPTION
**Pull request by kimchy**
Add the ability on top of status to know if a quorum of shards are active per index (so if there is at least one index without a quorum, it will be false).
 Also, allow to wait for it in the cluster health API.
 This change also centralizes all of these logic into a single enum class and makes sure the places where we check for it in our code base uses the same logic, with simple unit tests for it.

### CODE REVIEW

**Comment by javanna** [comment]
nice, this should solve #4755 I think?

**Comment by bleskes** [comment]
The change looks good and I like the new enum + isMet. However, since the write consistency is configurable per index for initial recovery and per write request, I think it makes sense to allow people to specify they consistency they want to wait for and also pick up on the index setting for the flag on the cluster health...

**Comment by kimchy** [comment]
@bleskes We don't allow to configure write consistency per index on "writes", we do allow to configure it per index for recovery, but those are completely 2 different things. For this reason, for the sake of simplicity, a simple "do we have quorum" (regardless of settings and such) makes sense.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java [hunk 1]
```diff
@@ -278,9 +291,7 @@ public void readFrom(StreamInput in) throws IOException {
         }
 
         numberOfInFlightFetch = in.readInt();
-        if (in.getVersion().onOrAfter(Version.V_1_7_0)) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java]: uh good! :)

File: core/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java [hunk 2]
```diff
@@ -191,6 +194,9 @@ private boolean prepareResponse(final ClusterHealthRequest request, final Cluste
         if (request.waitForStatus() != null && response.getStatus().value() <= request.waitForStatus().value()) {
             waitForCounter++;
         }
+        if (request.waitForQuorumActive() == true && response.isQuorumActive() == true) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java]: nit pick - we can skip the `== true` I think :)

File: core/src/main/java/org/elasticsearch/common/ConsistencyLevel.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public static ConsistencyLevel fromString(String value) {
+        if (value.equals("default")) {
+            return DEFAULT;
+        } else if (value.equals("one")) {
+            return ONE;
+        } else if (value.equals("quorum")) {
+            return QUORUM;
+        } else if (value.equals("all-1") || value.equals("full-1")) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/ConsistencyLevel.java]: can we please only have one value for this. there is no need to have more than one IMO

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/ConsistencyLevel.java]: oh damned it's BWC I guess...

File: core/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java [hunk 4]
```diff
@@ -191,6 +194,9 @@ private boolean prepareResponse(final ClusterHealthRequest request, final Cluste
...
+            waitForCounter++;
+        }
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java]: unrelated but can we add an assertion at the bottom of this method to ensure `waitForCounter <= waitFor`

**Comment by s1monw** [comment]
LGTM except of that I think we need to add some documentation for this feature. I left some minor comments

**Comment by s1monw** [comment]
>  For this reason, for the sake of simplicity, a simple "do we have quorum" (regardless of settings and such) makes sense.

+1 keep things simple and do not confuse people :)

**Comment by bleskes** [comment]
>  We don't allow to configure write consistency per index on "writes"

What I meant tis that we allow to configure it per request and also to configure the default using settings (`action.write_consistency`) .

> +1 keep things simple and do not confuse people :)

Agreed but if things are not consistent between APIs I think we end up causing confusion rather than making it better.

I'm not sure about the background for this change (feels like it's part of a bigger plan), but I want to repeat my suggestion from #4755 (@javanna you have the memory of an elephant - thanks for reminding me) :

> To me, yellow means “cluster is fully functional albeit not conforming to the required replication factor”. Red means that some operations can not be performed. With this definition in mind, I think we should signal an shard (and thus index & cluster) as red if one can not index into it due to the `write_consistency` settings (which also be `all` , and not `quorum`) 

If we do that, we can keep the current wait for logic as is and people can wait for yellow? 

Note: the suggestion should probably be adjusted to take `index.recovery.initial_shards` into account as well. For as long as we have it.. (different discussion)

**Comment by clintongormley** [comment]
> If we do that, we can keep the current wait for logic as is and people can wait for yellow?

I don't think we should change the existing meaning of yellow - it will just cause a lot of confusion as the existing meanings are well known and this change would be silent.

With the proposed change, users can wait for quorum or wait for green (if consistency is all).  They can also do this on a per-index basis.  I think that is clear, simple, and backwards compatible.

**Comment by kimchy** [comment]
ok, after a few discussions about it, I will break this pull request. First, the common way to parse what qualities to an active out of total is good, so I will keep this around.

Also, I want to use this opportunity to get away from the consistency level enum.

Last, we can add in a followup pull request anything that has to do with cluster health.

**Pull request closed without merging**